feat(replays): add ability to travel to previous and next replay

jerryzhou196 opened the pull request:
![playlist_demo](https://github.com/user-attachments/assets/cb8291a5-4e66-4a3e-9f41-3eb43e1da7af)

jerryzhou196 committed:
feat(replays): Enhance replay details navigation with start and end timestamps

Added functionality to display the start and end timestamps of replays in the ReplayTable and ReplayDetails components. This allows users to navigate between replays with context on their timing, improving the overall user experience. Updated ReplayDetailsUserBadge to include navigation buttons for previous and next replays.

jerryzhou196 committed:
ref(replays): Simplify ReplayDetailsUserBadge component. Removed unused props for next and previous replays in the ReplayDetailsUserBadge component, streamlining the code. Adjusted the loading placeholder height for better visual consistency. This change enhances the component's clarity and performance by eliminating unnecessary complexity

jerryzhou196 committed:
ref(replays): Update ReplayDetails and Breadcrumbs for improved navigation Refactored the ReplayDetails component to use 'undefined' instead of 'null' for next and previous replay checks, enhancing clarity. Updated ReplayDetailsPageBreadcrumbs to include navigation buttons for previous and next replays, improving user experience by allowing seamless navigation between replays. This change streamlines the code and enhances the overall functionality of the replay details view.

jerryzhou196 committed:
ref(replays): Update replay navigation icons and structure

Replaced the 'Next' icon in ReplayPlayPauseBar with 'IconChevron' for improved clarity. Refactored ReplayDetailsPageBreadcrumbs to utilize 'ButtonBar' for previous and next replay navigation buttons, enhancing the user interface and experience. This change streamlines the navigation between replays, making it more intuitive for users.

jerryzhou196 committed:
ref(replays): Simplify replay time handling in ReplayTable and related components

Refactored the ReplayTable component to streamline the handling of replay start times by removing the end timestamp logic. Updated related components to eliminate unnecessary props and improve clarity. This change enhances the overall performance and readability of the replay table, making it easier to manage replay data.

jerryzhou196 committed:
Merge branch 'master' into replay-carousal

jerryzhou196 committed:
Merge branch 'master' into replay-carousal

jerryzhou196 committed:
fix(replay-table): Enhance sorting functionality and update query parameters

- Imported `encodeSort` from `eventView` to handle sorting in the replay table.
- Updated the `ReplaySessionColumn` to utilize the new sorting mechanism without modifying the URL directly.
- Refactored query parameters in `ReplayDetails` to include `playlistStart` and improved handling of the `sort` parameter using `useUrlParams`.
- Exported `DEFAULT_SORT` from `useReplayTableSort` for broader accessibility.

jerryzhou196 committed:
Merge branch 'master' into replay-carousal

jerryzhou196 committed:
fix(replay): fix typo in DEFAULT_SORT

jerryzhou196 commented on static/app/components/replays/table/replayTable.tsx:
```diff
@@ -42,6 +43,17 @@ export default function ReplayTable({
+    const earliestReplayStartedAt = replays.reduce(
+      (acc: number, replay) => Math.min(replay.started_at?.getTime() ?? 0, acc),
+      Infinity
+    );
```
@billyvg I might be misunderstanding what you're saying, but currently it's doing up to earliestReplayStartedAt, so the list inside of details should remain the same? 

I did just push a fix because there was a bug in the `sort`, but I believe it works now. We can also hash this out tomorrow in-person

billyvg replied:
Right, that doesn't change my comment at all, earliest means oldest

billyvg commented on static/app/views/replays/details.tsx:
```diff
@@ -41,6 +47,50 @@ export default function ReplayDetails() {
+  const {data} = useApiQuery<{
+    data: ReplayListRecord[];
+    enabled: true;
+  }>(queryKey, {staleTime: 0});
```
What happens when we visit replay details from a non-replay index source? (e.g. from issue details)

billyvg commented on static/app/views/replays/details.tsx:
```diff
@@ -41,6 +47,50 @@ export default function ReplayDetails() {
+    },
+  });
+  const {getParamValue} = useUrlParams('sort');
+  const sortQuery = getParamValue();
```
isn't this the same as `query.sort`?

jerryzhou196 committed:
Merge remote-tracking branch 'origin/master' into replay-carousal

jerryzhou196 committed:
refactor(replayTable): Rename 'start' to 'end' for replay timestamps

Updated the replay table component to calculate and pass the latest replay timestamp as 'end' instead of 'start'. Adjusted related components and queries to reflect this change, ensuring consistency across the application.

jerryzhou196 commented on static/app/components/replays/table/useReplayTableSort.tsx:
```diff
@@ -12,7 +12,7 @@ interface Props {
 }
 
-const DEFAULT_SORT = {field: 'started_at', kind: 'asc'} as const;
+export const DEFAULT_SORT = {field: 'started_at', kind: 'desc'} as const;
```
I think `DEFAULT_SORT` is actually redundant in this hook to begin with, since we have a default value of `-started_at` earlier that is always set. But since we're exporting it now, I decided to change it to `desc` to match the earlier default value.

jerryzhou196 committed:
fix(replayTable): Update comment for sort query string handling

Clarified the comment in the ReplaySessionColumn regarding how EventView fetches the sort parameter from the URL, ensuring better understanding of the code's functionality.

jerryzhou196 committed:
feat(replayDetails): remove location.query

jerryzhou196 committed:
refactor(replayTable): Remove unused end timestamp and update query handling. Eliminated the calculation of the latest replay timestamp in the ReplayTable component, as it was no longer needed. Updated the ReplaySessionColumn to handle the event view query without the end parameter using exclusively the statsPeriod, ensuring cleaner code and improved clarity in the query string management.

jerryzhou196 committed:
added left margin to buttons

jerryzhou196 commented on static/app/components/replays/table/replayTableColumns.tsx:
```diff
@@ -522,18 +525,25 @@ export const ReplaySessionColumn: ReplayTableColumn = {
+          referrer,
+          ...eventView.generateQueryStringObject(),
+          playlistStart: start,
+          sort,
```
applied

jerryzhou196 committed:
address Billy's code review

jerryzhou196 committed:
feat(replayTable): Integrate EventView into ReplayTable and related components Added EventView support to the ReplayTable component, allowing for enhanced query handling and sorting. Updated the ReplayDetailsLinkColumn to utilize EventView for constructing query parameters, ensuring a more dynamic interaction with replay data. Adjusted the GroupReplays component to pass the eventView prop, improving overall functionality and user experience.

jerryzhou196 committed:
fix(replayDetails): Correct nextReplay calculation to handle edge cases. Updated the logic in the ReplayDetails component to ensure that the next replay is only accessed when the current index is valid, preventing potential out-of-bounds errors.

jerryzhou196 committed:
feat(replayTable): normalize eventView handling in ReplayTable and columns.

jerryzhou196 committed:
feat(replayTable): Enhance eventView query handling with stats period integration. Updated ReplayTableColumns to include a new function for generating query strings that incorporates the stats period, ensuring accurate date range handling in replay links.

jerryzhou196 committed:
Merge branch 'master' into replay-carousal

billyvg commented on static/app/views/replays/detail/header/replayDetailsPageBreadcrumbs.tsx:
```diff
@@ -89,21 +89,43 @@ export default function ReplayDetailsPageBreadcrumbs({readerResult}: Props) {
+            size="zero"
+            borderless
+            icon={<IconCopy size="xs" color="subText" />}
+          />
```
Let's do this instead, when not hovered, `visible: hidden`

jerryzhou196 replied:
nicee

jerryzhou196 committed:
removed changes for issueDetail Column view

jerryzhou196 committed:
undo replayController

jerryzhou196 committed:
Merge branch 'master' into replay-carousal

jerryzhou196 committed:
remove export

jerryzhou196 committed:
fix(replayTable): Refactor query generation to include playlist start and end dates when stats period is provided, improving date range handling in replay links.

jerryzhou196 committed:
feat(replayDetails): move hover state to only be in copy button for replay playlist

jerryzhou196 committed:
fix(replayTable): move sort fowarding back to index page

jerryzhou196 committed:
fix(replayTable): move sort fowarding back to index page

jerryzhou196 committed:
feat(replay): conditionally render if playlIstStart and playlistEnd are provided as args

jerryzhou196 committed:
fix(replay): refactor query handling in ReplaySessionColumn and ReplayDetails to ensure proper sorting and playlist date management

jerryzhou196 committed:
fix(replayDetails): conditionally render playlistView buttons

jerryzhou196 committed:
fix(replaySessionColumn): enhance query handling to include cursor from location and clarify sort field retrieval

jerryzhou196 commented:
Testing Notes: 
- Conditionally renders playlist buttons if feature flag is enabled 
- Playlists are enabled as long as `start`, `end` and `sort` are passed
- Playlists work for:
   - absolute / relative times
   - differing pages in the index page (first page, second page...) 
   - sorted columns in desc/asc order  
   - queries in search bar
   - most dead clicks / most rage clicks (disabled for these)
   - environment filter

jerryzhou196 committed:
fix(groupReplays.spec): update expected query, replacing statsPeriod with playlist_start and playlist_end and sort

jerryzhou196 committed:
Merge branch 'master' into replay-carousal

jerryzhou196 committed:
chore(replay): Update icon for next breadcrumb button from IconNext to IconChevron (#102985)

In prep for #102583, swap the `IconNext` logo with `IconChevron` to free
up for the replay playlist buttons.

jerryzhou196 committed:
Merge remote-tracking branch 'origin/master' into replay-carousal

jerryzhou196 committed:
fix(transactionReplays): update expected query to include playlistStart, playlistEnd, and sort parameters

jerryzhou196 committed:
feat(replay): introduce useReplayPlaylist hook for managing replay playlist queries

jerryzhou196 committed:
refactor(replayDetails): streamline breadcrumb component by moving replay navigation logic into ReplayDetailsPageBreadcrumbs

jerryzhou196 committed:
refactor(useReplayPlaylist): simplify query options by removing unnecessary object spread

jerryzhou196 committed:
feat(replay): integrate ReplayPlaylistProvider into ReplayDetails for improved context management

jerryzhou196 committed:
fix(replay): fix query property in transactions

jerryzhou196 committed:
Merge branch 'master' into fix-fetch-replay-list-hook

jerryzhou196 committed:
Merge branch 'fix-fetch-replay-list-hook' into replay-carousal

jerryzhou196 committed:
refactor(replay): update replay query handling and remove unused useReplayPlaylist hook

- Introduced REPLAY_LIST_FIELDS to streamline query construction in ReplaySessionColumn.
- Replaced useReplayPlaylist with useReplayList in ReplayDetailsProviders and related components for improved data handling.
- Updated type definitions to reflect changes in replay data structure.
- Removed the now obsolete useReplayPlaylist hook to clean up the codebase.

jerryzhou196 committed:
refactor(replay): update field mapping in ReplaySessionColumn for improved query handling

- Changed the field mapping in ReplaySessionColumn to use the full REPLAY_LIST_FIELDS array instead of splitting the fields. This enhances clarity and maintains consistency in query construction.

jerryzhou196 committed:
fix(replay): Add enabled option to useReplayList and adjust fetching logic

jerryzhou196 committed:
fix(replay): streamline saved query creation in useReplaysFromTransaction hook

jerryzhou196 committed:
Merge branch 'fix-fetch-replay-list-hook' into replay-carousal

jerryzhou196 committed:
Merge branch 'master' into replay-carousal

jerryzhou196 committed:
Merge branch 'master' into fix-fetch-replay-list-hook

jerryzhou196 committed:
fix(replay): set default value for enabled option in useReplayList hook

jerryzhou196 committed:
refactor(transactionReplays): simplify empty state test by removing redundant API call and using findByText for assertions

jerryzhou196 committed:
Merge branch 'master' into fix-fetch-replay-list-hook

jerryzhou196 committed:
Merge branch 'master' into fix-fetch-replay-list-hook

jerryzhou196 committed:
Merge branch 'fix-fetch-replay-list-hook' into replay-carousal

jerryzhou196 committed:
move replayPrev and replayNext icon to left of icons

jerryzhou196 committed:
Merge branch 'master' into fix-fetch-replay-list-hook

jerryzhou196 committed:
fix(replays): Ensure replay list fetching is correctly enabled

Updated the useReplayList hook to enforce fetching behavior based on the enabled flag. The flag is now required in the hook's options, and the fetching state is managed more consistently across components using the hook.

jerryzhou196 committed:
fix(replays): Simplify enabled flag condition for fetching replays

Updated the condition for enabling the replay list fetching in the useReplayList hook to directly check if the eventView.query is not an empty string. This change enhances clarity and maintains consistent fetching behavior.

jerryzhou196 committed:
Merge branch 'master' into fix-fetch-replay-list-hook

jerryzhou196 committed:
Merge branch 'master' into replay-carousal

jerryzhou196 committed:
Merge branch 'master' into replay-carousal

jerryzhou196 committed:
fix(replays): Update query parameters in tests and enable replay list fetching

Modified the expected query parameters in the GroupReplays and TransactionReplays tests to include additional fields for better data representation. Also, enabled the fetching of replay lists in the ReplayDetailsProviders component to ensure consistent behavior across the application.

jerryzhou196 committed:
fix(replays): Update replay list fetching condition

Modified the condition for enabling the replay list fetching in the useReplayList hook to ensure it only activates when both start and end times are defined, along with the presence of sorting parameters. This change improves the reliability of the replay list functionality.

jerryzhou196 committed:
refactor(replays): Enhance replay query handling and type definitions

Updated the replay query construction in ReplaySessionColumn to include playlist-specific parameters and improved the handling of location types in ReplayDetailsProviders. This refactor ensures that sorting and date range parameters are correctly integrated into the replay fetching logic, enhancing overall functionality and type safety.

jerryzhou196 committed:
Merge branch master into replay-carousal

jerryzhou196 committed:
fixed tests

jerryzhou196 commented on static/app/components/replays/table/replayTableColumns.tsx:
```diff
@@ -522,18 +523,40 @@ export const ReplaySessionColumn: ReplayTableColumn = {
+      query = {...query, ...{cursor: location.query.cursor}};
+    }
+
+    // Because the sort field is only generated if the corresponding fields is added, but we
```
reword to '...sort field is only generated if the corresponding fields are also specified in the URL'

jerryzhou196 commented on static/app/views/replays/detail/body/replayDetailsProviders.tsx:
```diff
@@ -38,6 +44,40 @@ export default function ReplayDetailsProviders({children, replay, projectSlug}:
+      newLocation.query.start = location.query.playlistStart;
+      newLocation.query.end = location.query.playlistEnd;
+    }
+    // TODO: import shared DEFAULT_SORT
```
do this

jerryzhou196 committed:
ref(replays): Update sorting logic to use shared constant for replay list

- Updated `replayDetailsProviders` to utilize `DEFAULT_REPLAY_LIST_SORT` for sorting logic.
- Improved comments in `replayTableColumns` for clarity on query handling.
- Added a blank line for better code readability in `useReplayTableSort`.

jerryzhou196 committed:
ref(replays): Refactor query handling in ReplayDetailsProviders

- Replaced `useLocation` with `useLocationQuery` for improved query parsing.
- Updated logic to construct a new query object based on playlist parameters.
- Integrated `useApiQuery` for fetching replay data, enhancing data handling.
- Mapped API response to replay records for better data structure.

jerryzhou196 committed:
ref(location): Change DefaultQuery type to local scope in useLocation.tsx

jerryzhou196 committed:
ref(replays): Simplify query object construction in ReplayDetailsProviders

- Destructured playlist parameters from the query object to streamline the creation of a new query.
- Enhanced readability and maintainability of the query handling logic.

jerryzhou196 commented on static/app/views/replays/detail/body/replayDetailsProviders.tsx:
```diff
@@ -38,6 +45,44 @@ export default function ReplayDetailsProviders({children, replay, projectSlug}:
+
+  const replays = useMemo(() => data?.data?.map(mapResponseToReplayRecord) ?? [], [data]);
+
   useLogReplayDataLoaded({projectId: replayRecord.project_id, replay});
```
good point, but replayIndexContainer uses the same pattern without any problems

jerryzhou196 committed:
ref(replays): Use useRef for initial location in ReplayDetailsPageBreadcrumbs

- Replaced direct usage of location with useRef to store the initial location.
- Updated query object construction to utilize the ref, enhancing consistency in query handling.

jerryzhou196 committed:
ref(replays): Refactor ReplayDetailsPageBreadcrumbs for improved styling

- Replaced `StyledFlex` with `Flex` for consistency in component usage.
- Introduced `StyledDiv` for better styling of the replay link.

jerryzhou196 committed:
ref(replays): Update queryReferrer in ReplayDetailsProviders for consistency

- Changed `queryReferrer` from 'replayList' to 'playlist' to align with updated naming conventions.
- This change enhances clarity in the query handling logic.

billyvg commented on static/app/components/replays/table/replayTableColumns.tsx:
```diff
@@ -522,18 +523,40 @@ export const ReplaySessionColumn: ReplayTableColumn = {
+      ...eventViewQuery,
+    };
+
+    if (!eventViewQuery.start && !eventViewQuery.end && typeof statsPeriod === 'string') {
```
This only checks for `statsPeriod`, now that we translate to `playlistStart`/`End`, we need to also account for `start` and `end`

billyvg commented on static/app/components/replays/table/replayTableColumns.tsx:
```diff
@@ -522,18 +523,40 @@ export const ReplaySessionColumn: ReplayTableColumn = {
+
+    const {statsPeriod, ...eventViewQuery} = eventView.generateQueryStringObject();
+
+    let query = {
```
[logaf](https://blog.danlew.net/2020/04/15/the-logaf-scale/) low: we use `query` in a lot of places, we could be a bit more specific and name this `detailsTabQuery`

billyvg commented on static/app/components/replays/table/replayTableColumns.tsx:
```diff
@@ -522,18 +523,40 @@ export const ReplaySessionColumn: ReplayTableColumn = {
+
+    if (!eventViewQuery.start && !eventViewQuery.end && typeof statsPeriod === 'string') {
+      const {start: playlistStart, end: playlistEnd} = parseStatsPeriod(statsPeriod);
+      query = {...query, ...{playlistStart, playlistEnd}};
```
m: no need to spread `query` - you can assign directly on the object

```suggestion
      query.playlistStart = playlistStart;
      query.playlistEnd = playlistEnd;
```

billyvg commented on static/app/components/replays/table/replayTableColumns.tsx:
```diff
@@ -522,18 +523,40 @@ export const ReplaySessionColumn: ReplayTableColumn = {
+    }
+
+    // Because the sort field is only generated if the corresponding fields are also in the URL, but we
+    // want to avoid dirtying the URL with fields, we manually add the sort field to the query here.
```
This also applies to `cursor` too right?

jerryzhou196 replied:
good point

billyvg commented on static/app/views/replays/detail/body/replayDetailsProviders.tsx:
```diff
@@ -38,6 +45,44 @@ export default function ReplayDetailsProviders({children, replay, projectSlug}:
+  const queryKey = useReplayListQueryKey({
+    options: {query: newQuery},
+    organization,
+    queryReferrer: 'playlist',
```
`replaysPlayList` maybe?

jerryzhou196 committed:
ref(replays): Enhance query handling in ReplaySessionColumn and ReplayDetailsProviders

- Updated query construction in `ReplaySessionColumn` to include `start` and `end` parameters for better playlist navigation.
- Renamed `queryReferrer` from 'playlist' to 'replaysPlayList' for consistency across components.
- Improved clarity in query handling by explicitly managing playlist parameters in `ReplayDetailsProviders`.
- Streamlined the query object to avoid confusion with regular parameters, enhancing maintainability.

jerryzhou196 committed:
ref(replays): Simplify sorting logic in ReplayDetailsProviders

- Updated query construction to use `DEFAULT_REPLAY_LIST_SORT` for sorting, ensuring consistent default behavior.
- Removed conditional check for `playlistSort` to streamline query handling logic.
- Enhanced the `enabled` flag in the query to focus on essential parameters, improving clarity in data fetching.

jerryzhou196 committed:
ref(replays): Rename StyledDiv to ShortId for clarity in ReplayDetailsPageBreadcrumbs

- Replaced `StyledDiv` with `ShortId` to better represent the component's purpose.
- Updated the component usage in the breadcrumb section to maintain consistency and improve readability.

jerryzhou196 committed:
ref(replays): Adjust sorting logic in ReplayDetailsProviders for clarity

- Modified the query sorting logic to use `DEFAULT_REPLAY_LIST_SORT` when `playlistSort` is an empty string, ensuring consistent default behavior.
- This change enhances the clarity of the query construction process in the component.

billyvg commented on static/app/views/replays/detail/body/replayDetailsProviders.tsx:
```diff
@@ -38,6 +46,45 @@ export default function ReplayDetailsProviders({children, replay, projectSlug}:
+    query.start = playlistStart;
+    query.end = playlistEnd;
+  }
+  query.sort = playlistSort === '' ? DEFAULT_REPLAY_LIST_SORT : playlistSort;
```
@jerryzhou196 Is cursor right? can `playlistSort` be undefined? if so we should probably do something like `playlistSort || DEFAULT`

jerryzhou196 committed:
ref(replays): Improve sorting logic in ReplayDetailsProviders for better clarity

- Updated the query sorting logic to handle cases where `playlistSort` is undefined or an empty string, ensuring consistent default behavior with `DEFAULT_REPLAY_LIST_SORT`.
- This change enhances the clarity and robustness of the query construction process.

jerryzhou196 committed:
ref(replays): Enhance parseStatsPeriod function to support UTC option

- Updated the parseStatsPeriod function to include an optional UTC parameter, allowing for more flexible date handling.
- Adjusted the ReplaySessionColumn to utilize the new UTC functionality when parsing the stats period, improving the accuracy of date calculations.

billyvg reviewed (approved):
🚢

jerryzhou196 commented on static/app/components/replays/table/replayTableColumns.tsx:
```diff
@@ -522,18 +524,48 @@ export const ReplaySessionColumn: ReplayTableColumn = {
+    }
+
+    // Because the sort and cursor field is only generated in EventView conditionally and we
+    // want to avoid dirtying the URL with fields, we manually add the them to the query here.
```
typo

jerryzhou196 committed:
ref(replays): Clarify comment in ReplaySessionColumn regarding query handling

- Updated comment in `ReplaySessionColumn` to improve clarity on the manual addition of fields to the query, enhancing code readability.
